src/main/scala/variant/Variant.scala

```diff
@@ -152,7 +152,10 @@ abstract class Variant private[variant] (
       !pawnsOnPromotionRank(board, color) &&
       !pawnsOnBackRank(board, color)
 
-  def valid(situation: Situation, strict: Boolean) = Color.all forall validSide(situation.board, strict)
+  def valid(situation: Situation, strict: Boolean) =
+    (Color.all forall validSide(situation.board, strict)) && hasValidCheckers(strict, situation)
+
+  protected def hasValidCheckers(strict: Boolean, situation: Situation): Boolean
```

Comment:
I think we shouldn't have this function in `Variant`. We're trying to keep Variant as lean as possible. So, we can do the same as before, adding checks on `valid` function for each variant if necessary.